rcl_yaml_param_parser/test/test_yaml_descriptor.cpp

```diff
@@ -0,0 +1,111 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include "osrf_testing_tools_cpp/scope_exit.hpp"
+#include "../src/impl/yaml_descriptor.h"
+#include "rcutils/allocator.h"
+#include "rcutils/strdup.h"
+
+#define TEST_DESCRIPTOR_COPY(field, tmp_field) \
+  do { \
+    SCOPED_TRACE("TEST_DESCRIPTOR_COPY " #field); \
+    rcl_param_descriptor_t src_descriptor{}; \
+    rcl_param_descriptor_t dest_descriptor{}; \
+    rcutils_allocator_t allocator = rcutils_get_default_allocator(); \
+    src_descriptor.field = &tmp_field; \
+    EXPECT_TRUE(rcl_yaml_descriptor_copy(&dest_descriptor, &src_descriptor, allocator)); \
+    ASSERT_NE(nullptr, dest_descriptor.field); \
+    EXPECT_EQ(*src_descriptor.field, *dest_descriptor.field); \
+    rcl_yaml_descriptor_fini(&dest_descriptor, allocator); \
+    src_descriptor.field = nullptr; \
+  } while (0)
+
+TEST(TestYamlDescriptor, copy_fini) {
+  rcl_param_descriptor_t descriptor{};
+  rcl_param_descriptor_t copy{};
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+
+  EXPECT_FALSE(rcl_yaml_descriptor_copy(nullptr, &descriptor, allocator));
+  EXPECT_FALSE(rcl_yaml_descriptor_copy(&copy, nullptr, allocator));
+
+  ASSERT_TRUE(rcl_yaml_descriptor_copy(&copy, &descriptor, allocator));
+
+  rcl_yaml_descriptor_fini(&copy, allocator);
+
+  // Check second fini works fine
+  rcl_yaml_descriptor_fini(&copy, allocator);
+
+  // Check fini with a nullptr doesn't crash.
+  rcl_yaml_descriptor_fini(nullptr, allocator);
+}
+
+TEST(TestYamlDescriptor, copy_fields) {
+  bool tmp_read_only = true;
+  TEST_DESCRIPTOR_COPY(read_only, tmp_read_only);
+
+  uint8_t tmp_type = 2;
+  TEST_DESCRIPTOR_COPY(type, tmp_type);
+
+  double tmp_min_value_double = -5.5;
+  TEST_DESCRIPTOR_COPY(min_value_double, tmp_min_value_double);
+
+  double tmp_max_value_double = 16.4;
+  TEST_DESCRIPTOR_COPY(max_value_double, tmp_max_value_double);
+
+  double tmp_step_double = 0.1;
+  TEST_DESCRIPTOR_COPY(step_double, tmp_step_double);
+
+  int64_t tmp_min_value_int = 1;
+  TEST_DESCRIPTOR_COPY(min_value_int, tmp_min_value_int);
+
+  int64_t tmp_max_value_int = 1001;
+  TEST_DESCRIPTOR_COPY(max_value_int, tmp_max_value_int);
+
+  int64_t tmp_step_int = 5;
+  TEST_DESCRIPTOR_COPY(step_int, tmp_step_int);
+}
+
+TEST(TestYamlDescriptor, copy_string_fields) {
+  // String version is slightly different and can't use the above macro
+  rcl_param_descriptor_t src_descriptor{};
+  rcl_param_descriptor_t dest_descriptor{};
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+
+  char * tmp_description = rcutils_strdup("param description", allocator);
```

Comment:
@bpwilcox nit: following above's suggestion, you may be able to merge this test into the one before by directly passing `rcutils_strdup()` and setting up a scope exit block to finalize `src_descriptor`.